Policies for Auth Token and Approve/Reject

cmc333333 opened the pull request:
Also separates our token login handling into a separate method.

Probably the most interesting bit here is that, when validating an auth token, we don't have access to a `User` (by definition). We can, however, treat the `params` hash as if it were a user, due to Pundit's flexible permissions model.

afeld commented on app/policies/proposal_policy.rb:
```diff
@@ -13,24 +14,32 @@ def initialize(user, proposal)
+    !@proposal.approvals.find_by(user: @user).nil?
+  end
+
+  def is_pending_approver?
```
Typically, methods that return booleans in Ruby aren't prefixed with `is_*`.

cmc333333 replied:
Okidok. I was trying to match the can_XXX names, but I can definitely strip off the is'es

afeld commented on spec/controllers/communicarts_controller_spec.rb:
```diff
@@ -130,24 +130,6 @@
-        token.update_attributes(used_at: 1.hour.ago)
-        bypass_rescue
-        expect { put 'approval_response', approval_params_with_token }.to raise_error(AuthenticationError)
-      end
```
I think it still might be useful to have a smoke test at the controller/request level, even if it just uses [mocks](https://relishapp.com/rspec/rspec-mocks/docs) to ensure that the appropriate policy is being called.

cmc333333 replied:
Can do.

afeld commented on app/controllers/communicarts_controller.rb:
```diff
@@ -19,69 +15,89 @@ def send_cart
+    else
+      self.user_access_errors(exception.query, exception.message)
+    end
   end
```
Would be nice to get this logic (basically all of the `protected` methods) out of the controller. Not a blocker, though.

cmc333333 replied:
I agree; the error messages are tied very tightly to the policy that failed, for example. I'll tinker with this a bit.

cmc333333 commented:
@afeld : updated to address your comments, thanks! Removed the TreePolicy as it was too confusing. Replaced it with an exception-based system. Policies are now written to raise exceptions (with specific error messages -- this binds error messages much closer to their policy). When the policies are called with a trailing `?`, that exception is caught, if raised, and `false` is returned instead.

Also removed all of the `is_` prefixes and added back the authorization smoke tests.

afeld commented on app/controllers/application_controller.rb:
```diff
@@ -10,24 +10,20 @@ class ApplicationController < ActionController::Base
-      super(record, q)  # have Pundit process each permission
+    # use the action as a default permission
+    query ||= ("can_" + params[:action].to_s + "!").to_sym
+    unless policy.public_send(query)
```
Passing user input to a `#send` is dangerous...think it's worth checking against a whitelist?

afeld replied:
...though maybe it's not as big of a deal in this case since it's prefixing (and suffixing) the method name.

cmc333333 replied:
I was under the assumption that `params[:action]` was a whitelist already?

All queries should be coming from code rather than a user. If we make a mistake and allow user input to get in, however, that user would only have access to the policy object, which is intentionally limited.

afeld replied:
Yeah, just tested...it gets overridden, even if that query param is present. Probably ok for now, but maybe worth flagging for a future security review.

afeld commented on app/policies/api_token_policy.rb:
```diff
@@ -0,0 +1,37 @@
+  end
+
+  def not_expired!
+    exists! && check(!@api_token.expires_at || @api_token.expires_at > Time.now,
```
Is checking for `!@api_token.expires_at` correct? Don't we want the opposite?

afeld replied:
Also, maybe move the latter check into an `ApiToken#expired?` method.

cmc333333 replied:
This duplicates existing logic: if there is no `expires_at` set on the token, we proceed. I assume this was a backwards compatibility thing.

Can definitely move that into `#expired?`

afeld replied:
> if there is no `expires_at` set on the token, we proceed.

That seems like a mistake, or a hand-wave-y thing we did to make our tests work. It *should* be present for all tokens. Storied:

https://www.pivotaltracker.com/story/show/92351222

afeld commented on app/policies/exception_policy.rb:
```diff
@@ -0,0 +1,35 @@
+    end
+    guard
+  end
+end
```
This class is complex enough that it could probably use some dedicated unit tests.

cmc333333 replied:
Can do.

afeld commented:
New approach works for me!

cmc333333 commented:
Updated to use an `expired?` method and add specs specifically for the ExceptionPolicy mixin
